File: native/ottaviano-ios/Sources/CoreModels/KDSLogic.swift
import Foundation

// Pure Kitchen Display ticket logic — the tone tiers, SLA meter, due countdown,
// channel tag, station grouping and allergen dedupe the KDS card renders. No
// SwiftUI, no colour: 1:1 with the web `src/lib/kds-prediction.ts` (ticketTone)
// + `src/core/kds/CoreKds.tsx` (dueLabel / slaPct) + `kds-board.ts` (grouping),
// so the native card draws byte-for-byte the same as the web board. Shared by
// `KDSTicket` (presentation) and the board KPIs (counts), exactly like the web
// shares these helpers — kept framework-free so it's trivially testable.

/// Live tone for a ticket (web `ticketTone`): ready → late → at-risk → warn →
/// firing/queued. Drives both the card accent and the due-text colour.
public enum KdsTone: Sendable, Equatable {
    case queued, firing, warn, risk, late, ready
}

public enum KDSClock {
    /// mm:ss for a positive (absolute) seconds value — the caller prepends any
    /// minus sign (web `fmtClock`). Used for the due timer + age KPIs.
    public static func clock(_ seconds: Double) -> String {
        let total = abs(Int(seconds.rounded()))
        return String(format: "%02d:%02d", total / 60, total % 60)
    }

    // Cached ISO parsers — `parseMs` runs on every periodic tick for every ticket
    // and ISO8601DateFormatter init is costly, so the formatters are reused.
    // `parseMs` is `public static` over `Sendable` types, so it can't assume a
    // single actor: a lock serializes the (uncontended, main-actor in practice)
    // access, since ISO8601DateFormatter mutation isn't guaranteed thread-safe.
    private static let parseLock = NSLock()
    nonisolated(unsafe) private static let isoFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    nonisolated(unsafe) private static let isoPlain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    /// ISO-8601 → ms epoch, tolerant of fractional seconds.
    public static func parseMs(_ iso: String) -> Double? {
        parseLock.lock()
        defer { parseLock.unlock() }
        if let d = isoFractional.date(from: iso) { return d.timeIntervalSince1970 * 1000 }
        if let d = isoPlain.date(from: iso) { return d.timeIntervalSince1970 * 1000 }
        return nil
    }
}

/// One station's lines on a ticket — `Identifiable` (by category) so the KDS
/// card can group multi-station tickets without index/keypath gymnastics.
public struct KdsItemGroup: Identifiable, Sendable {
    public var id: String { category }
    public let label: String
    public let category: String
    public let items: [OrderLine]
}

public extension Order {
    /// Elapsed-timer anchor (ms epoch) — paid time, else created (web
    /// `buildKdsTicket` paidAtMs).
    var paidAtMs: Double {
        KDSClock.parseMs(paidAt ?? createdAt) ?? Date().timeIntervalSince1970 * 1000
    }

    /// Promised-ready instant (ms epoch) from the prediction block, flattened.
    private var promisedMs: Double? { prediction.flatMap { $0.promisedReadyAtMs } }

    /// Live tone (web `ticketTone`). With a prediction block this is the
    /// SLA/at-risk model; without one (off-board single reads) it falls back to
    /// elapsed age so the card still reads sensibly.
    func kdsTone(nowMs: Double) -> KdsTone {
        if status == .ready { return .ready }
        guard prediction != nil else {
            let mins = max(0, (nowMs - paidAtMs) / 60_000)
            if mins >= 12 { return .late }
            if mins >= 5 { return .warn }
            return status == .confirmed ? .queued : .firing
        }
        if let promised = promisedMs {
            let slaRem = promised - nowMs
            if slaRem < 0 { return .late }
            if let predicted = prediction?.predictedReadyAtMs, predicted > promised { return .risk }
            if slaRem < 180_000 { return .warn }
        }
        return status == .confirmed ? .queued : .firing
    }

    /// Due text + its tone (web `dueLabel`): "done" when ready, "−mm:ss" when
    /// past the promise, the SLA countdown, or the predicted-ready countdown.
    func kdsDue(nowMs: Double) -> (text: String, tone: KdsTone) {
        let tone = kdsTone(nowMs: nowMs)
        if status == .ready { return ("done", tone) }
        if let promised = promisedMs {
            let slaRemSec = (promised - nowMs) / 1000
            if slaRemSec < 0 { return ("−" + KDSClock.clock(-slaRemSec), tone) }
            return (KDSClock.clock(slaRemSec), tone)
        }
        if let predicted = prediction?.predictedReadyAtMs {
            return (KDSClock.clock(max(0, (predicted - nowMs) / 1000)), tone)
        }
        return (KDSClock.clock(max(0, (nowMs - paidAtMs) / 1000)), tone)
    }

    /// Cook-time meter fill, 0 (fresh) → 1 (due) — web `slaPct` / 100.
    func slaFraction(nowMs: Double) -> Double {
        if status == .ready { return 1 }
        if let promised = promisedMs {
            let slaRemSec = (promised - nowMs) / 1000
            if slaRemSec < 0 { return 1 }
            let window = max(60, (promised - paidAtMs) / 1000)
            return min(1, max(0, 1 - slaRemSec / window))
        }
        let elapsed = max(0, (nowMs - paidAtMs) / 1000)
        let predRem = max(0, ((prediction?.predictedReadyAtMs ?? nowMs) - nowMs) / 1000)
        return min(0.95, elapsed / max(60, predRem + elapsed))
    }

    /// Past the promised-ready time and not yet plated — the board "Late" count.
    func isLate(nowMs: Double) -> Bool {
        guard status != .ready, let promised = promisedMs else { return false }
        return promised < nowMs
    }

    /// Model predicts a miss before it's actually late — the board "At risk" count.
    var isAtRisk: Bool { status != .ready && (prediction?.atRisk ?? false) }

    /// KDS channel chip (web `channelTag`): Dine-in (+ party size) / Delivery /
    /// Takeaway. Tolerates both `dine-in` (DTO) and `dine_in` (legacy) spellings.
    var channelTag: String {
        switch fulfillmentType {
        case "dine-in", "dine_in":
            return "Dine-in" + (partySize.map { " · \($0)p" } ?? "")
        case "delivery":
            return "Delivery"
        default:
            return "Takeaway"
        }
    }

    /// Canonical station order for grouping a multi-station ticket's lines
    /// (web `CATEGORY_ORDER`).
    static let kdsCategoryOrder = ["pizza", "pasta", "antipasti", "panini", "drinks", "desserts"]

    static func kdsCategoryLabel(_ c: String) -> String {
        c.isEmpty ? c : c.prefix(1).uppercased() + c.dropFirst()
    }

    /// Group lines by station in canonical order (web `groupItems`).
    func groupedItems() -> [KdsItemGroup] {
        var buckets: [String: [OrderLine]] = [:]
        var seen: [String] = []
        for it in items {
            let c = it.category ?? "other"
            if buckets[c] == nil { seen.append(c) }
            buckets[c, default: []].append(it)
        }
        func rank(_ c: String) -> Int { Self.kdsCategoryOrder.firstIndex(of: c) ?? 99 }
        return seen.sorted { rank($0) < rank($1) }
            .map { KdsItemGroup(label: Self.kdsCategoryLabel($0), category: $0, items: buckets[$0]!) }
    }

    /// Deduped allergens across the ticket's lines, in first-seen order — the
    /// KDS allergen callout (web `allergens` dedupe).
    var ticketAllergens: [String] {
        var seen = Set<String>()
        var out: [String] = []
        for it in items {
            for a in (it.allergens ?? []) where !a.isEmpty && !seen.contains(a) {
                seen.insert(a)
                out.append(a)
            }
        }
        return out
    }
}

File: native/ottaviano-rn/src/features/kds/kdsLogic.ts
import type { OrderDTO, OrderLineDTO, OrderStatus } from "@/api/types";
import { fmtClock, parseMs } from "@/lib/format";

/**
 * Pure Kitchen Display logic — the column model, station filters, status
 * progression, tone tiers, SLA meter, due countdown, channel tag, station
 * grouping and allergen dedupe the KDS card renders. 1:1 with the web
 * `src/lib/kds-prediction.ts` (ticketTone) + `src/core/kds/CoreKds.tsx`
 * (dueLabel / slaPct) + `kds-board.ts` (grouping) — and the Swift `KDSLogic`
 * port. No UI, no colour: the card maps a `KdsTone` to a palette colour. Operates
 * directly on the server-enriched `OrderDTO` (which already carries the
 * prediction block computed by `analyzeTruck`).
 */

export type KdsTone = "queued" | "firing" | "warn" | "risk" | "late" | "ready";

export const MENU_CATEGORY_LABELS: Record<string, string> = {
  pizza: "Pizza",
  pasta: "Pasta",
  antipasti: "Antipasti",
  panini: "Panini",
  drinks: "Drinks",
  desserts: "Desserts",
};

export const KDS_COLUMNS: { id: OrderStatus; label: string }[] = [
  { id: "confirmed", label: "New" },
  { id: "preparing", label: "Firing" },
  { id: "ready", label: "Ready · Expo" },
];

export const BUMP_LABEL: Partial<Record<OrderStatus, string>> = {
  confirmed: "Start firing",
  preparing: "Mark ready",
  ready: "Bump to pass",
};

export const STATION_FILTERS: { id: string; label: string }[] = [
  { id: "all", label: "All stations" },
  { id: "pizza", label: "Pizza" },
  { id: "pasta", label: "Pasta" },
  { id: "antipasti", label: "Antipasti" },
  { id: "panini", label: "Panini" },
  { id: "drinks", label: "Drinks" },
  { id: "desserts", label: "Desserts" },
];

export function nextStatus(current: OrderStatus): OrderStatus | null {
  if (current === "confirmed") return "preparing";
  if (current === "preparing") return "ready";
  if (current === "ready") return "completed";
  return null;
}

export function paidAtMs(o: OrderDTO): number {
  return parseMs(o.paidAt) ?? parseMs(o.createdAt) ?? Date.now();
}

/** Live tone (web `ticketTone`). With a prediction block this is the SLA/at-risk
 *  model; without one (off-board single reads) it falls back to elapsed age. */
export function toneFor(o: OrderDTO, nowMs: number): KdsTone {
  if (o.status === "ready") return "ready";
  const promised = o.prediction?.promisedReadyAtMs ?? null;
  if (!o.prediction) {
    const mins = Math.max(0, (nowMs - paidAtMs(o)) / 60000);
    if (mins >= 12) return "late";
    if (mins >= 5) return "warn";
    return o.status === "confirmed" ? "queued" : "firing";
  }
  if (promised !== null) {
    const slaRem = promised - nowMs;
    if (slaRem < 0) return "late";
    if (o.prediction.predictedReadyAtMs > promised) return "risk";
    if (slaRem < 180000) return "warn";
  }
  return o.status === "confirmed" ? "queued" : "firing";
}

/** Due text + tone (web `dueLabel`): "done" when ready, "−mm:ss" past promise,
 *  the SLA countdown, or the predicted-ready countdown. */
export function dueLabel(o: OrderDTO, nowMs: number): { text: string; tone: KdsTone } {
  const tone = toneFor(o, nowMs);
  if (o.status === "ready") return { text: "done", tone };
  const promised = o.prediction?.promisedReadyAtMs ?? null;
  if (promised !== null) {
    const slaRemSec = (promised - nowMs) / 1000;
    if (slaRemSec < 0) return { text: `−${fmtClock(-slaRemSec)}`, tone };
    return { text: fmtClock(slaRemSec), tone };
  }
  if (o.prediction) return { text: fmtClock(Math.max(0, (o.prediction.predictedReadyAtMs - nowMs) / 1000)), tone };
  return { text: fmtClock(Math.max(0, (nowMs - paidAtMs(o)) / 1000)), tone };
}

/** Cook-time meter fill 0→1 (web `slaPct` / 100). */
export function slaFraction(o: OrderDTO, nowMs: number): number {
  if (o.status === "ready") return 1;
  const promised = o.prediction?.promisedReadyAtMs ?? null;
  if (promised !== null) {
    const slaRemSec = (promised - nowMs) / 1000;
    if (slaRemSec < 0) return 1;
    const windowSec = Math.max(60, (promised - paidAtMs(o)) / 1000);
    return Math.min(1, Math.max(0, 1 - slaRemSec / windowSec));
  }
  const elapsed = Math.max(0, (nowMs - paidAtMs(o)) / 1000);
  const predRem = Math.max(0, ((o.prediction?.predictedReadyAtMs ?? nowMs) - nowMs) / 1000);
  return Math.min(0.95, elapsed / Math.max(60, predRem + elapsed));
}

export function isLate(o: OrderDTO, nowMs: number): boolean {
  if (o.status === "ready") return false;
  const promised = o.prediction?.promisedReadyAtMs ?? null;
  return promised !== null && promised < nowMs;
}

export function isAtRisk(o: OrderDTO): boolean {
  return o.status !== "ready" && (o.prediction?.atRisk ?? false);
}

export function channelTag(o: OrderDTO): string {
  if (o.fulfillmentType === "dine-in") return `Dine-in${o.partySize ? ` · ${o.partySize}p` : ""}`;
  if (o.fulfillmentType === "delivery") return "Delivery";
  return "Takeaway";
}

const CATEGORY_ORDER = ["pizza", "pasta", "antipasti", "panini", "drinks", "desserts"];
function catRank(c: string): number {
  const i = CATEGORY_ORDER.indexOf(c);
  return i < 0 ? 99 : i;
}
function catLabel(c: string): string {
  return MENU_CATEGORY_LABELS[c] ?? (c ? c[0].toUpperCase() + c.slice(1) : c);
}

/** Group lines by station in canonical order (web `groupItems`). */
export function groupItems(items: OrderLineDTO[]): { label: string; category: string; items: OrderLineDTO[] }[] {
  const buckets = new Map<string, OrderLineDTO[]>();
  for (const it of items) {
    const arr = buckets.get(it.category) ?? [];
    arr.push(it);
    buckets.set(it.category, arr);
  }
  return [...buckets.entries()]
    .sort((a, b) => catRank(a[0]) - catRank(b[0]))
    .map(([cat, arr]) => ({ label: catLabel(cat), category: cat, items: arr }));
}

export function ticketAllergens(o: OrderDTO): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const it of o.items) for (const a of it.allergens) if (a && !seen.has(a)) (seen.add(a), out.push(a));
  return out;
}

/** Group tickets into board columns with the station filter, oldest-first. */
export function groupByColumn(orders: OrderDTO[], station: string): Map<OrderStatus, OrderDTO[]> {
  const map = new Map<OrderStatus, OrderDTO[]>();
  for (const col of KDS_COLUMNS) map.set(col.id, []);
  for (const o of orders) {
    if (station !== "all" && !o.items.some((i) => i.category === station)) continue;
    map.get(o.status)?.push(o);
  }
  for (const arr of map.values()) arr.sort((a, b) => paidAtMs(a) - paidAtMs(b));
  return map;
}
